---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```
# bbcRNA: An R package for facilitating RNA-Seq analysis

<!-- badges: start -->
[![Travis build status](https://travis-ci.org/vari-bbc/bbcRNA.svg?branch=master)](https://travis-ci.org/vari-bbc/bbcRNA)
[![Codecov test coverage](https://codecov.io/gh/vari-bbc/bbcRNA/branch/master/graph/badge.svg)](https://codecov.io/gh/vari-bbc/bbcRNA?branch=master)
<!-- badges: end -->

The goal of bbcRNA is to facilitate RNA-seq analysis, starting from counts and
ending with differentially expressed gene lists and enerichment analyses. The
major feature are classes to store different parts of the RNA-Seq analysis and
wrapper functions for running popular tools such as EdgeR and ClusterProfiler.
See [the bbcRNA website](https://vari-bbc.github.io/bbcRNA/) for a built
vignette.

## Installation

You can install bbcRNA from [GitHub](https://github.com/vari-bbc) with:

``` r
if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")
  
devtools::install_github("vari-bbc/bbcRNA")
```
## New classes defined in the bbcRNA package

New classes have been defined to enable more convenient storage of different
parts of a DE analysis.

### The BbcSE class

The `BbcSE` class extends SummarizedExperiment. Additional slots are:

1. `aln_metrics`: a matrix containing alignment metrics for samples.
2. `edger` : a `BbcEdgeR` object.
3. `deseq2`: For storing DESeqDataSet and DESeqResults; not implemented yet.

### The BbcEdgeR class

Slots are:

1. `dgelist`: a `DGEList` object.
2. `de_results`: a list. First element is always a `DGEGLM` object. The others
are edgeR result objects.
3. `norm_cts`: a `SummarizedExperiment` object for storing normalized counts.

## Usage

```{r}
# attach the package
library(bbcRNA)

# run the BbcSE constructor without any data to show the structure 
BbcSE()

# run the BbcEdgeR constructor without any data to show the structure 
BbcEdgeR()

```
    
