---
title: "Using the bbcRNA package for RNA-seq analysis"
author: Kin Lau
output: 
  html_document:
        toc: true
        toc_depth: 2
        toc_float: true
vignette: >
  %\VignetteIndexEntry{rnaseq_from_star}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(

)
```

In this vignette, we demonstrate how to use the bbcRNA package to run an RNA-seq
analysis starting from gene read count files produced by the STAR aligner.

# Install bbcRNA if not installed
```{r}
if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")
if (!requireNamespace("bbcRNA", quietly = TRUE))
  devtools::install_github("vari-bbc/bbcRNA")
```


# Load packages

```{r setup}
library(bbcRNA)
library(SummarizedExperiment)
library(org.Mm.eg.db) # load the org.db for your organism

```

# Get counts

```{r get_counts}
tcell <- star_to_mat(dir = "../inst/extdata/tcell/", rgx = "^[^_]+_[^_]+", column = 2)

head(tcell)
```

# make GRanges

This step is slow (several minutes to finish). It can be skipped if the genomic 
intervals of the genes are not of interest.

```{r granges}
# txdb <- GenomicFeatures::makeTxDbFromGFF("~/Desktop/gencode.vM13.annotation.gtf",
#                                         format="gtf")
# granges <- GenomicFeatures::genes(txdb)
```


# Make BbcSE object

The granges parameter for the BbcSE constructor is optional. If not provided,
dummy GRanges data will be used. SummarizedExperiment can be coerced to 
RangedSummarizedExperiment without GRanges data, but DEFormat does not work
with RangedSummarizedExperiment objects with missing data for rowRanges.
 
```{r make_obj}
# bbc_obj <- BbcSE(counts = tcell, granges = granges)
bbc_obj <- BbcSE(counts = tcell)

# show more information about the BbcSE class
getClassDef(class(bbc_obj))

bbc_obj
```

# Get sample meta data

```{r sample_meta}
col_meta <- read_col_meta("../inst/extdata/tcell/meta.txt")

# Add column meta data. May implement a function to add column meta data
# eventually
colData(bbc_obj) <- cbind(colData(bbc_obj), col_meta[colnames(bbc_obj), ])

colData(bbc_obj)
```

# Get mapping rates

```{r get_mapping_rates}
aln_rates <- read_star_map_rates(dir = "../inst/extdata/tcell/", 
                                 rgx = "^[^_]+_[^_]+")

# store the mapping rates in the BbcSE object
aln_rates(bbc_obj) <- aln_rates

# aln_rates is the name of both the setter and the getter function for aln_rates
aln_rates(bbc_obj)
```

# Change the column (sample) names

```{r change_sample_names}
# add new column in colData by concatenating two genotype and rep columns
colData(bbc_obj)$new_sample <- paste0(colData(bbc_obj)$genotype, 
                                      colData(bbc_obj)$rep)

# store a copy of the original column names
colData(bbc_obj)$orig_sample <- colnames(bbc_obj)

# change the colnames to new_sample
colnames(bbc_obj) <- colData(bbc_obj)$new_sample
```

# Plot alignment rates

```{r plot_mapping}
plot_aln_rates(x = bbc_obj, type = "uniq_aln_reads", facet_by="genotype", fill="genotype") + ggplot2::theme(legend.position = "none")

plot_aln_rates(x = bbc_obj, type = "uniq_map_rate", facet_by="genotype", fill="genotype") + ggplot2::theme(legend.position = "none")
```

# Add gene symbols
```{r add_gene_syms}
bbc_obj <- ens2sym(bbc_obj, org.Mm.eg.db)

rowData(bbc_obj)
```

# make DGEList object
```{r make_DGEList}
# by default, low expression genes filtered out, normalization factors
# calculated, normalized counts calculated
bbc_obj <- makeDGEList(bbc_obj, group="genotype")

# what's in the edger slot now?
str(edger(bbc_obj))

# Number of rows in the SE
nrow(bbc_obj)

# Number of genes in edger$DGEList
nrow(dgelist(edger(bbc_obj)))
```
# PCA
```{r plot_pca}
pca_plots <- plot_PCA(bbc_obj, color_by="genotype", shape_by="genotype")
pca_plots[[1]]
pca_plots[[2]]
```

# edgeR DE workflow

```{r find_degs}
bbc_obj <- findDEGs(bbc_obj, 
                    de_pkg = "edger",
                    test = "glmQLFTest", 
                    design = "~0+genotype",
                    contrasts = list(c("genotype", "mut", "WT")))

```


# Session info
```{r}
sessionInfo()
```


