---
title: "RNA-seq analysis starting from gene counts from STAR aligner"
output: 
  BiocStyle::html_document:
        toc: true
        toc_depth: 2
        toc_float: true
vignette: >
  %\VignetteIndexEntry{rnaseq_from_star}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(

)
```

# Load packages

```{r setup}
library(bbcRNA)
library(SummarizedExperiment)
```

# Get counts

```{r get_counts}
tcell <- star_to_mat(dir = "../inst/extdata/tcell/", rgx = "^[^_]+_[^_]+", column = 2)

head(tcell)
```

# make GRanges

This step is slow (several minutes to finish). It can be skipped if the genomic 
intervals of the genes are not of interest.

```{r granges}
txdb <- GenomicFeatures::makeTxDbFromGFF("~/Desktop/gencode.vM13.annotation.gtf",
                                         format="gtf")
granges <- GenomicFeatures::genes(txdb)
```


# Make BbcSE object

The granges parameter for the BbcSE constructor is optional. If not provided,
dummy GRanges data will be used. SummarizedExperiment can be coerced to 
RangedSummarizedExperiment without GRanges data, but DEFormat does not work
with RangedSummarizedExperiment objects with missing data for rowRanges.
 
```{r make_obj}
bbc_obj <- BbcSE(counts = tcell, granges = granges)

# show more information about the BbcSE class
getClassDef(class(bbc_obj))

metadata(bbc_obj)[[1]] <- DEFormats::DGEList(bbc_obj)

bbc_obj
```

# Get mapping rates

```{r get_mapping_rates}
aln_rates <- read_star_map_rates(dir = "../inst/extdata/tcell/", 
                                 rgx = "^[^_]+_[^_]+")

# store the mapping rates in the BbcSE object
colData(bbc_obj) <- as(aln_rates, "DataFrame") 

# colData is the name of both the setter and the getter function for column data
head(colData(bbc_obj))
```

# Session info
```{r}
sessionInfo()
```


