---
title: "RNA-seq analysis starting from gene counts from STAR aligner"
output: 
  BiocStyle::html_document:
        toc: true
        toc_depth: 2
        toc_float: true
vignette: >
  %\VignetteIndexEntry{rnaseq_from_star}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(

)
```

# Load packages

```{r setup}
library(bbcRNA)
library(SummarizedExperiment)
library(org.Mm.eg.db) # load the org.db for your organism

```

# Get counts

```{r get_counts}
tcell <- star_to_mat(dir = "../inst/extdata/tcell/", rgx = "^[^_]+_[^_]+", column = 2)

head(tcell)
```

# make GRanges

This step is slow (several minutes to finish). It can be skipped if the genomic 
intervals of the genes are not of interest.

```{r granges}
# txdb <- GenomicFeatures::makeTxDbFromGFF("~/Desktop/gencode.vM13.annotation.gtf",
#                                         format="gtf")
# granges <- GenomicFeatures::genes(txdb)
```


# Make BbcSE object

The granges parameter for the BbcSE constructor is optional. If not provided,
dummy GRanges data will be used. SummarizedExperiment can be coerced to 
RangedSummarizedExperiment without GRanges data, but DEFormat does not work
with RangedSummarizedExperiment objects with missing data for rowRanges.
 
```{r make_obj}
# bbc_obj <- BbcSE(counts = tcell, granges = granges)
bbc_obj <- BbcSE(counts = tcell)

# show more information about the BbcSE class
getClassDef(class(bbc_obj))

bbc_obj
```

# Get sample meta data

```{r sample_meta}
col_meta <- read_col_meta("../inst/extdata/tcell/meta.txt")

# may implement a function to add column meta data eventually
colData(bbc_obj) <- cbind(colData(bbc_obj), col_meta[colnames(bbc_obj), ])

colData(bbc_obj)
```

# Get mapping rates

```{r get_mapping_rates}
aln_rates <- read_star_map_rates(dir = "../inst/extdata/tcell/", 
                                 rgx = "^[^_]+_[^_]+")

# store the mapping rates in the BbcSE object
aln_rates(bbc_obj) <- aln_rates

# aln_rates is the name of both the setter and the getter function for aln_rates
aln_rates(bbc_obj)
```

# Plot alignment rates

```{r plot_mapping}
plot_aln_rates(x = bbc_obj, type = "uniq_aln_reads", facet_by="genotype", fill="genotype") + ggplot2::theme(legend.position = "none")

plot_aln_rates(x = bbc_obj, type = "uniq_map_rate", facet_by="genotype", fill="genotype") + ggplot2::theme(legend.position = "none")
```

# Add gene symbols
```{r add_gene_syms}
bbc_obj <- ens2sym(bbc_obj, org.Mm.eg.db)

rowData(bbc_obj)
```

# edgeR
```{r make_DGEList}
# metadata(bbc_obj)$edger[[1]] <- DEFormats::DGEList(bbc_obj)
bbc_obj <- makeDGEList(bbc_obj, group="genotype")

metadata(bbc_obj)$edger
```


# Session info
```{r}
sessionInfo()
```


