<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run hypergeometric test — run_hypergeometric • bbcRNA</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/darkly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Run hypergeometric test — run_hypergeometric" />

<meta property="og:description" content="Run hypergeometric test" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bbcRNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rnaseq_from_star.html">Using the bbcRNA package for RNA-seq analysis</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run hypergeometric test</h1>
    
    <div class="hidden name"><code>run_hypergeometric.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Run hypergeometric test</p>
    
    </div>

    <pre class="usage"><span class='fu'>run_hypergeometric</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S4 method for data.frame</span>
<span class='fu'>run_hypergeometric</span>(<span class='no'>x</span>, <span class='kw'>gene_set</span> <span class='kw'>=</span> <span class='st'>"reactome"</span>,
  <span class='no'>organism</span>, <span class='no'>orgDb</span>, <span class='kw'>split_by_lfc_dir</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>padj_cutoff</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>lfc_cutoff</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='no'>...</span>)

<span class='co'># S4 method for BbcSE</span>
<span class='fu'>run_hypergeometric</span>(<span class='no'>x</span>, <span class='kw'>de_pkg</span> <span class='kw'>=</span> <span class='st'>"edger"</span>,
  <span class='kw'>contrast_names</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>split_by_lfc_dir</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>A BbcSE object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Passed to enrichKEGG, enrichGO, enrichPathway or enricher. See
each function for possible arguments.</p></td>
    </tr>
    <tr>
      <th>gene_set</th>
      <td><p>one of "kegg", "reactome", "H", "C1", "C2", "C3", "C4", "C5",
"C6", "C7"</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>For compatibility with dowstream tools, the organism is
specified differently depending on the desired gene set. For KEGG, use the
three-letter code according to
http://www.genome.jp/kegg/catalog/org_list.html. For Reactome, possible
values are ‘celegans’, ‘fly’, ‘human’, ‘mouse’, ‘rat’, ‘yeast’ and
‘zebrafish’. For msigdb, run 'msigdbr::msigdbr_show_species()'.</p></td>
    </tr>
    <tr>
      <th>orgDb</th>
      <td><p>OrgDB object for your organism</p></td>
    </tr>
    <tr>
      <th>split_by_lfc_dir</th>
      <td><p>logical. Whether DE genes should be split into up and
down-regulated</p></td>
    </tr>
    <tr>
      <th>padj_cutoff</th>
      <td><p>numeric. Cutoff for adjusted PValue for DE.</p></td>
    </tr>
    <tr>
      <th>lfc_cutoff</th>
      <td><p>numeric. Cutoff for LFC (absolute value) for DE.</p></td>
    </tr>
    <tr>
      <th>de_pkg</th>
      <td><p>"edger" or "deseq2". Only used if x is a BbcSE</p></td>
    </tr>
    <tr>
      <th>contrast_names</th>
      <td><p>character value or vector for the contrast(s) of
interest. See name(de_results(edger(x))). If left to default "" value then
all contrasts will be processed.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of gseaResult objects or one gseaResult object</p>
    
    <h2 class="hasAnchor" id="methods-by-class-"><a class="anchor" href="#methods-by-class-"></a>Methods (by class)</h2>

    
<ul>
<li><p><code>data.frame</code>: x is a dataframe with gene name (entrez) as
col1, log2FC as col2 and adjusted PValue as col3. Function will filter by
LFC and PValue for DE genes, and optionally will split the DE genes into up
and down-regulated based on LFC. The gene universe are all the genes in the
dataframe. Returns a list of enrichResults for up and down-regulaed genes
or just all the DE genes without regard for direction.</p></li>
<li><p><code>BbcSE</code>: Extract out the gene, LFC and adjusted PValue.
Run run_hypergeometric for each result object (contrast).</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://rdrr.io/pkg/clusterProfiler/man/enrichKEGG.html'>enrichKEGG</a>
  <a href='https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html'>enrichGO</a> <a href='https://rdrr.io/pkg/ReactomePA/man/enrichPathway.html'>enrichPathway</a>
  <a href='https://rdrr.io/pkg/clusterProfiler/man/enricher.html'>enricher</a></code></p></div>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#methods-by-class-">Methods (by class)</a></li>

      <li><a href="#see-also">See also</a></li>
          </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Kin Lau, Dean Pettinga, Ian Beddows, Scott Givan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


